### PR TITLE
[huggingface_pytorch][NEURON][build] Huggingface Neuron inference DLC

### PR DESCRIPTION
**Pull request by philschmid**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Description

This PR will add Hugging Face Inference Neuron DLCs for PyTorch. 

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [x] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [x] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [x] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [x] (If applicable) I've documented below the tests I've run on the DLC image
- [x] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [x] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [x] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by philschmid** [comment]
This is connected/related to 
* https://github.com/aws/deep-learning-containers/pull/1577

@arjkesh @saimidu

File: huggingface/pytorch/buildspec-neuron.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      target: neuron-entrypoint.py
+    config:
+      source: ../../build_artifacts/inference/config.properties
+      target: config.properties
+    neuron-monitor:
+      source: ../../build_artifacts/inference/neuron-monitor.sh
+      target: neuron-monitor.sh
+
+images:
+  BuildHuggingFacePytorchCpuPy38InferenceDockerImage:
```
Comment (aws-vrnatham) [inline huggingface/pytorch/buildspec-neuron.yml:28]: Please remove Cpu reference in this file and add Neuron if applicable.

File: huggingface/pytorch/inference/docker/1.9/py3/sdk1.16.1/Dockerfile.neuron [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+LABEL maintainer="Amazon AI"
+LABEL dlc_major_version="1"
+
+# Specify accept-bind-to-port LABEL for inference pipelines to use SAGEMAKER_BIND_TO_PORT
+# https://docs.aws.amazon.com/sagemaker/latest/dg/inference-pipeline-real-time.html
+LABEL com.amazonaws.sagemaker.capabilities.accept-bind-to-port=true
+
+ARG MMS_VERSION=1.1.4
+ARG PYTHON=python3
+ARG PYTHON_VERSION=3.8.10
```
Comment (aws-vrnatham) [inline huggingface/pytorch/inference/docker/1.9/py3/sdk1.16.1/Dockerfile.neuron:12]: Do we need python version 3.8 for HF. The recommended/tested python versions with torch-neuron are 3.6 & 3.7. If there is no hard requirement for 3.8 then can we use 3.7 like in [neuron](https://github.com/aws/deep-learning-containers/blob/master/pytorch/inference/docker/1.9/py3/sdk1.16.1/Dockerfile.neuron)

**Review by aws-vrnatham — COMMENTED** [review]
In test_pre_release.py for  test_framework_and_neuron_sdk_version please add the versions expected in NEURON_VERSION_MANIFEST for huggingface_pytorch so that this can pass

**Comment by arjkesh** [comment]
@philschmid - it looks like the tests are failing because "neuron" is not a supported processor type in the huggingface inference conftest. I would make the changes but it looks like this PR was made from your fork's master branch, so don't want to mess anything up on your setup

**Comment by philschmid** [comment]
> @philschmid - it looks like the tests are failing because "neuron" is not a supported processor type in the huggingface inference conftest. I would make the changes but it looks like this PR was made from your fork's master branch, so don't want to mess anything up on your setup

Feel free to make the changes and push. You are not messing up anything

File: huggingface/pytorch/buildspec-neuron.yml (hunk 1 above)
Comment (philschmid) [reply huggingface/pytorch/buildspec-neuron.yml]: Thanks for noticing! changed it to `BuildNeuronHuggingFacePytorchPy38InferenceDockerImage`

File: huggingface/pytorch/inference/docker/1.9/py3/sdk1.16.1/Dockerfile.neuron (hunk 2 above)
Comment (philschmid) [reply huggingface/pytorch/inference/docker/1.9/py3/sdk1.16.1/Dockerfile.neuron]: Changed to Python3.7. 
I wouldn't go with Python3.6 since it is deprecated End of Month

File: test/sagemaker_tests/huggingface/inference/integration/sagemaker/test_neuron_hosting.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+
+from ...integration import model_dir, pt_neuron_model,pt_neuron_script
+from ...integration.sagemaker.timeout import timeout_and_delete_endpoint
+from .... import invoke_pytorch_helper_function
+
+
+@pytest.mark.model("tiny-distilbert")
+@pytest.mark.processor("neuron")
+@pytest.mark.neuron_test
```
Comment (arjkesh) [inline test/sagemaker_tests/huggingface/inference/integration/sagemaker/test_neuron_hosting.py:32]: in order to use this marker you'll need to update conftest to do something with this marker

Comment (philschmid) [reply test/sagemaker_tests/huggingface/inference/integration/sagemaker/test_neuron_hosting.py:32]: I thought it existed already. I basically copy-pasted the contest from PyTorch for neuron

**Review by aws-vrnatham — APPROVED** [review]
I am seeing test only for sagemaker and assuming that HF inference containers can only be used in the SM environment and not EC2/EKS/ECS

File: huggingface/pytorch/inference/docker/1.9/py3/sdk1.16.1/Dockerfile.neuron [hunk 4]
```diff
@@ -0,0 +1,139 @@
+FROM ubuntu:20.04
+
+LABEL maintainer="Amazon AI"
+LABEL dlc_major_version="1"
+
+# Specify accept-bind-to-port LABEL for inference pipelines to use SAGEMAKER_BIND_TO_PORT
+# https://docs.aws.amazon.com/sagemaker/latest/dg/inference-pipeline-real-time.html
+LABEL com.amazonaws.sagemaker.capabilities.accept-bind-to-port=true
+
+ARG MMS_VERSION=1.1.5
```
Comment (arjkesh) [inline huggingface/pytorch/inference/docker/1.9/py3/sdk1.16.1/Dockerfile.neuron:10]: @philschmid do we need MMS for this release?

File: huggingface/pytorch/inference/docker/1.9/py3/sdk1.16.1/Dockerfile.neuron [hunk 5]
```diff
@@ -7,7 +7,7 @@ LABEL dlc_major_version="1"
 # https://docs.aws.amazon.com/sagemaker/latest/dg/inference-pipeline-real-time.html
 LABEL com.amazonaws.sagemaker.capabilities.accept-bind-to-port=true
 
-ARG MMS_VERSION=1.1.5
+ARG TS_VERSION=0.4.0
```
Comment (philschmid) [inline huggingface/pytorch/inference/docker/1.9/py3/sdk1.16.1/Dockerfile.neuron:10]: Ehm why did you add torchserve? We are not using Torchserve for Inference? The Inference Toolkit is built on top of MMS for both frameworks. https://github.com/aws/sagemaker-huggingface-inference-toolkit

File: huggingface/pytorch/inference/docker/1.9/py3/sdk1.16.1/Dockerfile.neuron (hunk 4 above)
Comment (philschmid) [reply huggingface/pytorch/inference/docker/1.9/py3/sdk1.16.1/Dockerfile.neuron]: What do you mean if we need it?

File: huggingface/pytorch/inference/docker/1.9/py3/sdk1.16.1/Dockerfile.neuron (hunk 5 above)
Comment (aws-vrnatham) [reply huggingface/pytorch/inference/docker/1.9/py3/sdk1.16.1/Dockerfile.neuron:10]: oh.. Will add mms back. Thanks.

Comment (arjkesh) [reply huggingface/pytorch/inference/docker/1.9/py3/sdk1.16.1/Dockerfile.neuron]: Can you set MMS to 1.1.7 when reverting?

**Comment by philschmid** [comment]
What changes do we need to make, to get this in?

**Comment by arjkesh** [comment]
lgtm, let's revert toml

**Comment by arjkesh** [comment]
@aws-vrnatham does the testing look sufficient to you?

**Review by arjkesh — APPROVED** [review]

**Comment by aws-vrnatham** [comment]
> @aws-vrnatham does the testing look sufficient to you?
Yes. Looks good to me

**Review by ahsan-z-khan — APPROVED** [review]

**Comment by philschmid** [comment]
Thank you guys for your support on this 🚀
